File: apps/quickink/android/app/src/main/java/app/quickink/mobile/features/scan/VoiceNoteTranscriptionPane.kt
/*
 * VoiceNoteTranscriptionPane.kt
 *
 * Second phase of the scan capture flow. Mounted by
 * [ScanCaptureSurface] after [VoiceNoteCapturePane] reports a
 * just-recorded voice note. Shows the live transcript in an
 * editable text field while the background transcribe pass lands,
 * and exposes a single "Continue to review" CTA that saves any
 * edits before advancing to [ScanReviewScreen].
 *
 * Lifecycle:
 *   - Observes `voice_notes` for the in-flight capture and picks
 *     out the row by id. While `transcription` is null we render a
 *     "Transcribing your note…" placeholder; once it lands we seed
 *     the editor exactly once so subsequent user edits stick.
 *   - On Continue: if the user changed the text we call
 *     [VoiceNoteRepository.setTranscription] to persist the edit,
 *     then invoke `onContinue`.
 *   - Close in the header deletes the just-recorded voice note,
 *     then returns to the add-context recorder phase so the user
 *     can re-record before review.
 */

package app.quickink.mobile.features.scan

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.WindowInsets
import androidx.compose.foundation.layout.asPaddingValues
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.statusBars
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Close
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.TextRange
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.KeyboardCapitalization
import androidx.compose.ui.text.input.TextFieldValue
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import app.quickink.mobile.QuickInkApp
import app.quickink.mobile.data.voicenote.VoiceNoteRepository
import app.quickink.mobile.ui.theme.LocalQuickInkColors
import app.quickink.mobile.ui.theme.LocalQuickInkTypography
import app.quickink.mobile.ui.theme.QuickInkRadius
import app.quickink.mobile.ui.theme.QuickInkSpacing
import app.releaf.mobile.ui.theme.AppSpacing
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.launch

@Composable
fun VoiceNoteTranscriptionPane(
    captureId: String,
    voiceNoteId: String,
    showCancelButton: Boolean = true,
    placeContinueBelowEditor: Boolean = false,
    continueLabel: String = "Continue to review",
    onContinue: () -> Unit,
    onCancel: () -> Unit,
) {
    val colors  = LocalQuickInkColors.current
    val type    = LocalQuickInkTypography.current
    val context = LocalContext.current
    val app     = context.applicationContext as QuickInkApp
    val repo    = remember(app) { VoiceNoteRepository(app.database.voiceNoteDao()) }
    val scope   = rememberCoroutineScope()

    // Observe just this row by filtering the per-capture flow.
    // Adding a dedicated `observeById` would be cleaner but isn't
    // worth a new DAO method for one call site.
    val rowFlow = remember(captureId, voiceNoteId) {
        repo.observeForCapture(captureId).map { rows ->
            rows.firstOrNull { it.id == voiceNoteId }
        }
    }
    val row by rowFlow.collectAsState(initial = null)

    var editedValue by remember(voiceNoteId) { mutableStateOf(TextFieldValue("")) }
    var seeded by remember(voiceNoteId) { mutableStateOf(false) }
    var userEdited by remember(voiceNoteId) { mutableStateOf(false) }
    var canceling by remember(voiceNoteId) { mutableStateOf(false) }

    // Seed the editor once when the transcript lands so the user's
    // subsequent edits aren't clobbered by later observation ticks
    // (e.g. if a sync push echoes the same value back).
    LaunchedEffect(row?.transcription) {
        val text = row?.transcription
        if (!seeded && !text.isNullOrEmpty() && !userEdited) {
            editedValue = TextFieldValue(
                text      = text,
                selection = TextRange(text.length),
            )
            seeded = true
        }
    }

    val transcribing = row?.transcription.isNullOrBlank() && !seeded

    val statusBarTop = WindowInsets.statusBars.asPaddingValues().calculateTopPadding()

    Column(
        modifier = Modifier
            .fillMaxSize()
            .background(colors.bg)
            .padding(top = statusBarTop),
    ) {
        // Header
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(
                    start  = QuickInkSpacing.s4,
                    end    = QuickInkSpacing.s4,
                    top    = QuickInkSpacing.s6,
                    bottom = QuickInkSpacing.s2,
                ),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            if (showCancelButton) {
                Box(
                    modifier = Modifier
                        .size(32.dp)
                        .clip(CircleShape)
                        .background(Color.White.copy(alpha = 0.8f))
                        .clickable(enabled = !canceling) {
                            canceling = true
                            scope.launch {
                                runCatching { repo.softDelete(voiceNoteId) }
                                onCancel()
                            }
                        },
                    contentAlignment = Alignment.Center,
                ) {
                    Icon(
                        imageVector       = Icons.Filled.Close,
                        contentDescription = "Close",
                        tint              = colors.inkSoft,
                        modifier          = Modifier.size(14.dp),
                    )
                }
            } else {
                Spacer(modifier = Modifier.size(50.dp))
            }
            Spacer(modifier = Modifier.weight(1f))
            Column(horizontalAlignment = Alignment.CenterHorizontally) {
                Text(
                    text  = "VOICE NOTE",
                    style = type.label.copy(
                        fontSize     = 10.sp,
                        fontWeight   = FontWeight.Bold,
                        letterSpacing = 1.2.sp,
                    ),
                    color = colors.muted,
                )
                Text(
                    text  = "Review transcript",
                    style = type.body.copy(
                        fontSize   = 15.sp,
                        fontWeight = FontWeight.SemiBold,
                    ),
                    color = colors.ink,
                )
            }
            Spacer(modifier = Modifier.weight(1f))
            // Match the capture pane's header layout (50dp gutter).
            Spacer(modifier = Modifier.size(50.dp))
        }

        // Body
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(horizontal = QuickInkSpacing.s4)
                .padding(top = QuickInkSpacing.s3),
        ) {
            Text(
                text  = if (transcribing) "Transcribing your note…" else "Edit the transcript before sending it to review.",
                style = type.meta,
                color = colors.muted,
            )

            Spacer(modifier = Modifier.size(QuickInkSpacing.s3))

            val editorShape = RoundedCornerShape(QuickInkRadius.md)
            Box(
                modifier = Modifier
                    .fillMaxWidth()
                    .clip(editorShape)
                    .background(Color.White.copy(alpha = 0.85f), editorShape)
                    .border(1.dp, colors.border, editorShape)
                    .padding(horizontal = QuickInkSpacing.s3, vertical = QuickInkSpacing.s3),
            ) {
                BasicTextField(
                    value         = editedValue,
                    onValueChange = {
                        editedValue = it
                        userEdited = true
                    },
                    textStyle     = type.body.copy(color = colors.ink, fontSize = 14.sp),
                    cursorBrush   = SolidColor(colors.accent),
                    keyboardOptions = KeyboardOptions(
                        capitalization = KeyboardCapitalization.Sentences,
                    ),
                    minLines = 8,
                    maxLines = 14,
                    modifier      = Modifier
                        .fillMaxWidth()
                        .heightIn(min = 160.dp, max = 320.dp),
                    decorationBox = { inner ->
                        if (editedValue.text.isEmpty()) {
                            Text(
                                text  = if (transcribing) "Listening for your words…"
                                        else "No transcript was generated. You can type notes here.",
                                style = type.body.copy(fontSize = 14.sp),
                                color = colors.muted,
                            )
                        }
                        inner()
                    },
                )
            }
            if (placeContinueBelowEditor) {
                Spacer(modifier = Modifier.size(QuickInkSpacing.s3))
                ContinueReviewButton(
                    label   = continueLabel,
                    onClick = {
                        persistTranscriptAndContinue(
                            editedText  = editedValue.text,
                            original    = row?.transcription.orEmpty(),
                            voiceNoteId = voiceNoteId,
                            source      = row?.transcriptionSource ?: "manual",
                            repo        = repo,
                            scope       = scope,
                            onContinue  = onContinue,
                        )
                    },
                )
            }
        }

        if (!placeContinueBelowEditor) {
            Spacer(modifier = Modifier.weight(1f))

            // Bottom bar — single "Continue to review" CTA.
            Box(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(
                        start  = AppSpacing.s5,
                        end    = AppSpacing.s5,
                        bottom = AppSpacing.s5,
                    ),
            ) {
                ContinueReviewButton(
                    label   = continueLabel,
                    onClick = {
                        persistTranscriptAndContinue(
                            editedText  = editedValue.text,
                            original    = row?.transcription.orEmpty(),
                            voiceNoteId = voiceNoteId,
                            source      = row?.transcriptionSource ?: "manual",
                            repo        = repo,
                            scope       = scope,
                            onContinue  = onContinue,
                        )
                    },
                )
            }
        }
    }
}

@Composable
private fun ContinueReviewButton(
    label: String,
    onClick: () -> Unit,
) {
    val colors = LocalQuickInkColors.current
    val type   = LocalQuickInkTypography.current
    Box(
        modifier = Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(999.dp))
            .background(colors.accent)
            .clickable(onClick = onClick)
            .padding(vertical = AppSpacing.s3),
        contentAlignment = Alignment.Center,
    ) {
        Text(
            text  = label,
            style = type.body.copy(fontWeight = FontWeight.SemiBold),
            color = colors.textOnAccent,
        )
    }
}

private fun persistTranscriptAndContinue(
    editedText: String,
    original: String,
    voiceNoteId: String,
    source: String,
    repo: VoiceNoteRepository,
    scope: kotlinx.coroutines.CoroutineScope,
    onContinue: () -> Unit,
) {
    if (editedText != original) {
        scope.launch {
            runCatching {
                repo.setTranscription(
                    id     = voiceNoteId,
                    text   = editedText.ifBlank { null },
                    // Preserve whichever backend produced the
                    // original draft; fall back to "manual" for a
                    // pure user-typed transcript.
                    source = source,
                )
            }
            onContinue()
        }
    } else {
        onContinue()
    }
}

File: apps/quickink/ios/QuickInk/Scan/VoiceNoteTranscriptionPane.swift
/*
 * VoiceNoteTranscriptionPane.swift
 *
 * Pre-review transcript editor for a freshly recorded voice note,
 * including the audio note extracted from video capture. The user can
 * edit the recognized text before moving into the normal scan review
 * surface, where the transcript feeds suggested tags.
 */

import Combine
import SwiftUI
import ReleafCoreDesignSystem

struct VoiceNoteTranscriptionPane: View {

    let captureId: String
    let voiceNoteId: String
    let showCancelButton: Bool
    let placeContinueBelowEditor: Bool
    let continueLabel: String
    let onContinue: () -> Void
    let onCancel: () -> Void

    @StateObject private var model: VoiceNoteTranscriptionPaneModel
    @State private var editedText: String = ""
    @State private var seeded: Bool = false
    @State private var userEdited: Bool = false
    @State private var canceling: Bool = false

    init(
        captureId: String,
        voiceNoteId: String,
        showCancelButton: Bool = true,
        placeContinueBelowEditor: Bool = false,
        continueLabel: String = "Continue to review",
        onContinue: @escaping () -> Void,
        onCancel: @escaping () -> Void
    ) {
        self.captureId = captureId
        self.voiceNoteId = voiceNoteId
        self.showCancelButton = showCancelButton
        self.placeContinueBelowEditor = placeContinueBelowEditor
        self.continueLabel = continueLabel
        self.onContinue = onContinue
        self.onCancel = onCancel
        _model = StateObject(
            wrappedValue: VoiceNoteTranscriptionPaneModel(
                captureId: captureId,
                voiceNoteId: voiceNoteId
            )
        )
    }

    private var transcribing: Bool {
        ((model.note?.transcription ?? "").isEmpty) && !seeded
    }

    private var editorBinding: Binding<String> {
        Binding(
            get: { editedText },
            set: { next in
                editedText = next
                userEdited = true
            }
        )
    }

    var body: some View {
        VStack(spacing: 0) {
            header
                .padding(.horizontal, QuickInkSpacing.s4)
                .padding(.top, QuickInkSpacing.s6)

            bodyContent
                .padding(.horizontal, QuickInkSpacing.s4)
                .padding(.top, QuickInkSpacing.s3)

            if placeContinueBelowEditor {
                Spacer()
            } else {
                Spacer(minLength: QuickInkSpacing.s5)
                bottomBar
                    .padding(.horizontal, QuickInkSpacing.s5)
                    .padding(.bottom, QuickInkSpacing.s5)
            }
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .background(QuickInkColors.bg.ignoresSafeArea())
        .task { model.start() }
        .onReceive(model.$note) { note in
            seedTranscript(note?.transcription)
        }
    }

    // MARK: Header

    private var header: some View {
        HStack(alignment: .center) {
            if showCancelButton {
                Button(action: cancelTranscriptReview) {
                    Image(systemName: "xmark")
                        .font(.system(size: 14, weight: .semibold))
                        .foregroundStyle(QuickInkColors.inkSoft)
                        .padding(8)
                        .background(Color.white.opacity(0.8), in: Circle())
                }
                .buttonStyle(.plain)
                .disabled(canceling)
                .accessibilityLabel("Close")
            } else {
                Color.clear.frame(width: 50, height: 1)
            }

            Spacer()

            VStack(spacing: 2) {
                Text("VOICE NOTE")
                    .font(.system(size: 10, weight: .bold))
                    .tracking(1.2)
                    .foregroundStyle(QuickInkColors.muted)
                Text("Review transcript")
                    .font(.system(size: 15, weight: .semibold))
                    .foregroundStyle(QuickInkColors.ink)
            }

            Spacer()
            Color.clear.frame(width: 50, height: 1)
        }
    }

    // MARK: Body

    private var bodyContent: some View {
        VStack(alignment: .leading, spacing: QuickInkSpacing.s3) {
            Text(transcribing ? "Transcribing your note..." : "Edit the transcript before sending it to review.")
                .font(QuickInkText.caption)
                .foregroundStyle(QuickInkColors.muted)

            editorSurface

            if placeContinueBelowEditor {
                continueButton
            }
        }
    }

    private var editorSurface: some View {
        ZStack(alignment: .topLeading) {
            TextEditor(text: editorBinding)
                .font(QuickInkText.body)
                .foregroundStyle(QuickInkColors.ink)
                .scrollContentBackground(.hidden)
                .textInputAutocapitalization(.sentences)
                .autocorrectionDisabled(false)
                .padding(QuickInkSpacing.s2)
                .frame(minHeight: 180, maxHeight: 320)

            if editedText.isEmpty {
                Text(transcribing ? "Listening for your words..." : "No transcript was generated. You can type notes here.")
                    .font(QuickInkText.body)
                    .foregroundStyle(QuickInkColors.muted)
                    .padding(QuickInkSpacing.s3)
                    .allowsHitTesting(false)
            }
        }
        .background(Color.white.opacity(0.85))
        .clipShape(RoundedRectangle(cornerRadius: QuickInkRadius.md, style: .continuous))
        .overlay(
            RoundedRectangle(cornerRadius: QuickInkRadius.md, style: .continuous)
                .stroke(QuickInkColors.border, lineWidth: 1)
        )
    }

    // MARK: Continue

    private var bottomBar: some View {
        continueButton
    }

    private var continueButton: some View {
        Button(action: continueReview) {
            Text(continueLabel)
                .font(AppText.body)
                .fontWeight(.semibold)
                .foregroundStyle(QuickInkColors.textOnAccent)
                .frame(maxWidth: .infinity)
                .padding(.vertical, AppSpacing.s3)
                .background(QuickInkColors.accent)
                .clipShape(Capsule())
        }
        .buttonStyle(.plain)
    }

    private func seedTranscript(_ text: String?) {
        guard !seeded, !userEdited else { return }
        guard let text, !text.isEmpty else { return }
        editedText = text
        seeded = true
    }

    private func continueReview() {
        let original = model.note?.transcription ?? ""
        let source = model.note?.transcriptionSource ?? "manual"
        guard editedText != original else {
            onContinue()
            return
        }
        Task {
            await model.saveTranscript(text: editedText, source: source)
            await MainActor.run { onContinue() }
        }
    }

    private func cancelTranscriptReview() {
        canceling = true
        Task {
            await model.deleteVoiceNote()
            await MainActor.run { onCancel() }
        }
    }
}

@MainActor
private final class VoiceNoteTranscriptionPaneModel: ObservableObject {

    @Published var note: VoiceNoteEntity? = nil

    private let captureId: String
    private let voiceNoteId: String
    private let repo: VoiceNoteRepository
    private var cancellable: AnyCancellable? = nil

    init(
        captureId: String,
        voiceNoteId: String,
        repo: VoiceNoteRepository = VoiceNoteRepository()
    ) {
        self.captureId = captureId
        self.voiceNoteId = voiceNoteId
        self.repo = repo
    }

    func start() {
        guard cancellable == nil else { return }
        cancellable = repo.observeForCapture(captureId)
            .receive(on: DispatchQueue.main)
            .sink(
                receiveCompletion: { _ in },
                receiveValue: { [weak self] rows in
                    guard let self else { return }
                    note = rows.first { $0.id == self.voiceNoteId }
                }
            )
    }

    func saveTranscript(text: String, source: String) async {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        try? await repo.setTranscription(
            id: voiceNoteId,
            transcription: trimmed.isEmpty ? nil : text,
            source: trimmed.isEmpty ? nil : source
        )
    }

    func deleteVoiceNote() async {
        try? await repo.softDelete(id: voiceNoteId)
    }
}
